blackbird_python/blackbird/tests/test_program.py

```diff
@@ -403,3 +405,90 @@ def test_serialize_operation_multiple_arrays(self):
             """
         )
         assert res == expected
+
+
+class TestProgramIntegration:
+
+    def test_template_with_arrays(self):
+        """Test templates can be initialized with parameters inside of arrays"""
+        template=dedent(
+            """\
+            name template_tdm
+            version 1.0
+
+            float array one[1, 2] =
+                {p_one}
+
+            float array two[2, 3] =
+                1, {p_two}, 3
+                1, 2, {p_three}
+
+            int four = {p_four}
+
+            for int i in 0:2
+                Sgate(two[i], 0) | 1
+                BSgate(one[i]) | (1, 0)
+                MeasureHomodyne(four) | 0
+            """
+        )
+
+        bb = loads(template)
+
+        assert bb.parameters == {
+            'p_one_0_0', 'p_one_0_1', 'p_two', 'p_three', 'p_four'
+        }
+
+        assert bb.is_template()
+
+        bb2 = bb(p_one=[[11, 12]], p_two=22, p_three=33, p_four=44)
+        assert not bb2.parameters
+        assert not bb2.is_template()
+
+        expected = {'one': np.array([[11, 12]]), 'two': np.array([[1, 22, 3], [1, 2, 33]]), 'four': 44}
+
+        assert np.all(np.all(v == expected[k]) for k, v in bb2._var.items())
+
+        assert bb2.operations[0] == {'op': 'Sgate', 'args': [1, 0], 'kwargs': {}, 'modes': [1]}
```

Comment:
Is the indexing done using the flattened array for the first element? :thinking: So `array([ 1, 22,  3,  1,  2, 33])[0] == 1`? :thinking:

Reply:
Yes, that's exactly how it works! This was done earlier as a simple way of handling indexing (and is detailed in [the new docs PR](https://github.com/XanaduAI/blackbird/pull/28)). :slightly_smiling_face: